MSC2290: Separate Endpoints for Threepid Binding (#2290)
matrix-org/matrix-doc: proposals/2290-separate-threepid-bind-hs.md

the threepid would be either added to the homeserver, or added to the
homeserver and bound to the identity server simultaneously.

Now, consider the fact that the identity server used in this process is
provided by the user, using the endpoint's `id_server` parameter. If the user were
to supply a malicious identity server that would immediately answer "yes" to
any threepid validation, then the user could add any threepid to their
account on the homeserver (which is likely not something homeserver admins want).

To be clear, this is not a long-standing security issue. It is not a problem
in any released version of Synapse, as Synapse keeps a list of "trusted
identity servers" that acts a whitelist for what identity servers a user can
specify.

The concept of this whitelist is being removed in this MSC however, as part
of lessening the reliance of homeservers on identity servers. This cannot be
done while the homeserver is still trusting an identity server for validation
of threepids. If the endpoints are split, the homeserver will handle the
validation of threepids being added to user accounts, and identity servers
will validate threepids being bound to themselves.

## Proposal

> To solve this problem, two new endpoints will be added to the Client Server

Reviewer: I might be missing it, but: is it ever spelt out what the HS is meant to do when these endpoints are hit? It's kinda implied by reference to the existing endpoint, but it seems a bit unclear to me.

Author: I've stated after each call what the accompanying server should have done internally. Hopefully that makes things clearer.

> To solve this problem, two new endpoints will be added to the Client Server

Reviewer: Relatedly: my main question is how hitting /bind on the HS differs from hitting the IS directly.

Author: The HS will just proxy this request. However, it will remember that the bind occurred, and when it comes time to unbind all of a user's threepids (during account deactivation or when they explicitly request it), then the homeserver is aware of what binds occurred where and can carry out the unbinds on the user's behalf.

But other than that it is simply just proxying the request.

# Separate Endpoints for Binding Threepids

On the Client Server API there is currently a single endpoint for binding a
threepid (an email or a phone number): [POST

> /_matrix/client/r0/account/3pid](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-3pid).

Author: I believe I've made it easier to read now? Please check.

> /_matrix/client/r0/account/3pid](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-3pid).

Reviewer 2: yup - thank you!

> # Separate Endpoints for Binding Threepids

Reviewer 2: The intro duplicates the proposal and is a bit verbose - it can be shortened by not explaining what the proposal is about to do and just setting the scene for why the changes are being made. 

The intro is already desynchronized from the proposal itself: it does not mention that `bind` is disappearing.

On the Client Server API there is currently a single endpoint for binding a
threepid (an email or a phone number): [POST

> /account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid).

Reviewer 2: please link to a specific version if at all possible. If the endpoint can't be linked to a specific version, mention that it's the endpoint scheduled for r0.6.0

identity server. This terminology will be used throughout the rest of this
proposal.

Typically, when using the `/account/3pid` endpoint, the identity server
handles the verification -- either by sending an email to an email address,
or a SMS message to a phone number. Once completed, the homeserver will check
with the identity server that verification had indeed happened, and if so,
the threepid would be either added to the homeserver, or added to the
homeserver and bound to the identity server simultaneously.

Now, consider the fact that the identity server used in this process is
provided by the user, using the endpoint's `id_server` parameter. If the user were
to supply a malicious identity server that would immediately answer "yes" to
any threepid validation, then the user could add any threepid to their
account on the homeserver (which is likely not something homeserver admins want).

To be clear, this is not a long-standing security issue. It is not a problem
in any released version of Synapse, as Synapse keeps a list of "trusted
identity servers" that acts a whitelist for what identity servers a user can
specify.

> Synapse is soon to lose this whitelist however, as part of lessening the

Reviewer 2: while great that Synapse wants to separate itself from identity servers, it's not a problem for the spec to deal with. Please present the feature without the reliance on Synapse - we wouldn't merge a feature that was implementation-specific because that implementation wanted it.

reliance of homeservers on identity servers. This cannot be done while the
homeserver is still trusting an identity server for validation of threepids.
If the endpoints are split, the homeserver will handle the validation of
threepids being added to user accounts, and identity servers will validate
threepids being bound to themselves.

> To solve this problem, we propose adding two new endpoints. `POST

Reviewer 2: we? 
```suggestion
To solve this problem, it is proposed to add two new endpoints. `POST
```

/account/3pid`) will be deprecated.

One may question why clients don't just contact an identity server directly
to bind a threepid, bypassing the implications of binding through a
homeserver. While this will work, binds should still occur through a
homeserver such that the homeserver can keep track of which binds were made,
which is important when a user wishes to deactivate their account (and remove
all of their bindings made on different identity servers).

A verification could occur on an identity server, which could then tell the
homeserver that a validation happened, but then there are security
considerations about how to authenticate an identity server in that instance
(and prevent people pretending to be identity servers and telling homeservers
about hundreds of fake threepid additions to a user's account).

This MSC obseletes
[MSC2229](https://github.com/matrix-org/matrix-doc/pull/2229), which dealt
with changing the rules of the `bind` flag on the original endpoint. Since
that endpoint is being deprecated, the MSC is no longer relevant.

## Proposal

Two new endpoints will be added to the Client Server API: `POST

> /account/3pid/bind` and `POST /account/3pid/add`. Both will require

Reviewer 2: auth + can be rate limited presumably. I know we've done a bad job of determining what can be rate limited thus far, but at some point we should stop being bad - this MSC is very much an arbitrary point for me to raise it.

authentication. The request parameters of `POST /account/3pid/bind` are the
same as [POST
/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid),
minus the `bind` flag. The request parameters of `POST /account/3pid/add`
will simply consist of a JSON body containing `client_secret` and `sid`.

An example of binding a threepid to an identity server with this new endpoint
is as follows:

First the client must request the threepid be validated by its chosen identity server.

```
POST https://identity.server/_matrix/identity/v2/validate/email/requestToken

{
    "client_secret": "don'tT3ll",
    "email": "bob@example.com",
    "send_attempt": 1
}
```

Once an email has been sent, the user clicks the link in the email, which
notifies the identity server that the email has been verified.

Next, the client completes the bind by calling the new endpoint on the homeserver:

```

> POST https://home.server/_matrix/client/r0/account/3pid/bind

Reviewer 3: Does the new HS `bind` endpoint require the 3PID to already have been added to the HS account previously? (I assume the answer is no, but the MSC doesn't appear to specify.)

Author: No, as the IS doesn't allow that. I'll add it in though, thanks.

{

>     "three_pid_creds": {

Reviewer 2: let's fix https://github.com/matrix-org/matrix-doc/issues/2220 while we're here and converge on exactly one. I don't think we need to be terse here and can say "threepid_credentials"

Reviewer 2: or better yet just not wrap the object

notifies the homeserver that the threepid has been verified.

The client then sends a request to the endpoint on the homeserver to add 
the threepid to a user's account.

```
POST https://home.server/_matrix/client/r0/account/3pid/add

{
    "sid": "abc123987",
    "client_secret": "don'tT3ll"
}
```

The threepid has now been added to the user's account.

To achieve the above flows, some changes need to be made to existing
endpoints. This MSC requests that the `id_server` and `id_access_token`
parameters be removed from the Client-Server API's [POST
/account/3pid/email/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-email-requesttoken)
and [POST
/account/3pid/msisdn/requestToken](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-msisdn-requesttoken)
endpoints, as these endpoints are now only intended for the homeserver to
send validation requests from.

Additionally, the [POST
/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid)
endpoint is deprecated as the two new endpoints replace its functionality.

> The `bind` endpoint will also be removed, with the endpoint functioning as if

Reviewer 3: What is that `bind` endpoint you are removing? (I thought this MSC was adding the bind endpoint...)

Author: whoops, meant "parameter", not endpoint. Thanks.

> The `bind` parameter will also be removed, with the endpoint functioning as if

Reviewer 2: ```suggestion
The `bind` parameter is to be removed, with the endpoint functioning as if
```

`bind` was `false`. Allowing an endpoint to add a threepid to both the
identity server and homeserver at the same time requires one to trust the
other, which is the exact behaviour we're trying to eliminate. Doing this
also helps backward compatibility, as explained below.

> This MSC also requests that the text "It is imperative that the homeserver

Reviewer 2: MSCs are demands by nature (proposals go forward with recommendations leading to action, not listing things that don't require action) - if this is meant to be actionable, please phrase it as such. My recommendation is:

> The text "It is imperative...that it trusts" is to be removed from all parts of the spec, as the homeserver should no longer need to trust any identity servers.

keep a list of trusted Identity Servers and only proxies to those that it
trusts." be removed from all parts of the spec, as the homeserver should no
longer need to trust any identity servers.

## Tradeoffs

It may be possible to reduce the two calls per flow into a single endpoint,
but the current asynchronous approach makes it easy for a client to send a
request, go offline, have the threepid be validated, and then come online
again to finalize the validation afterwards.

> ## Backwards compatibility

Reviewer 2: Does this need an unstable feature flag so that clients wiring up these new APIs can detect their presence?

Reviewer 3: Hmm, I suppose an unstable flag might make the client work slightly simpler... Without a flag, you would need to probe the new endpoints and check for 404s to detect support. If it were just a single API call, 404 detection could be okay, but in this case, these APIs are part of a multiple step flow of requesting tokens, sending to a service, etc. and the new endpoints are not the first step, so it means the 404 detection would be an extra check before starting the entire flow.

Author: I've switched it to use a new unstable flag, `m.separate_add_and_bind`.

Old matrix clients will continue to use the `/account/3pid` endpoint. This
MSC removes the `bind` parameter and forces `/account/3pid` calls to act as
if `bind` was set to `false`. Old clients will still be able to add 3pids to
the homeserver, but not bind to the identity server. New homeservers must
ignore any `id_server` information passed to this endpoint.

New matrix clients running with old homeservers should try their desired
endpoint (either `/account/3pid/add` or `/account/3pid/bind`) and on
receiving a HTTP `404` error code, should either attempt to use
`/account/3pid` with the `bind` parameter or give up, at their discretion.

## Security considerations

Reducing the homeserver's trust in identity servers should be a boost to
security and improve decentralisation in the Matrix ecosystem to boot.

Caution should be taken for homeserver developers to be sure not to continue

> to use user-provided identity servers for any sensitive tasks once it removes

Reviewer 2: I have no idea what this entails - what sensitive tasks is this referencing? why am I supposed to be careful but allow any ID server to go through?

the concept of a trusted identity server list.

> ## Conclusion

Reviewer 2: I feel like this whole section can be removed from this MSC (and the template, for that matter)

# Separate Endpoints for Binding Threepids

On the Client Server API there is currently a single endpoint for binding a
threepid (an email or a phone number): [POST
/account/3pid](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid).
Depending on whether the `bind` flag is `true` or `false`, the threepid will

> be bound to either a user's account on the homeserver, or both the homeserver

Reviewer 3: In an attempt to preserve some shreds on sanity with 3PIDs, I have been using the verb "bind" _only_ for the case of identity server. For adding a 3PID to an HS account, I use the verb "add". Using separate verbs in this way makes the concepts easier to distinguish.

I think this MSC would be easier to read if it followed the same convention, as opposed to have two meanings of binding depending on the service.

Reviewer 4: definitely agreed. (plus we should define that terminology up front)

homeserver, or the homeserver and identity server simultaneously).

Now, consider the fact that the identity server used in this process is
provided by the user, using the endpoint's `id_server` parameter. If the user were
to supply a malicious identity server that would immediately answer "yes" to
any threepid validation, then the user could add any threepid to their
account on the homeserver (which is likely not something homeserver admins want).

To be clear, this is not a long-standing security issue. It is not a problem
in any released version of Synapse, as Synapse keeps a list of "trusted
identity servers" that acts a whitelist for what identity servers a user can
specify.

Synapse is soon to lose this whitelist however, as part of lessening the
reliance of homeservers on identity servers. This cannot be done while the
homeserver is still trusting an identity server for validation of threepids.
If the endpoints are split, the homeserver will handle the validation of
threepids being added to user accounts, and identity servers will validate
threepids being added to their own database.

To solve this problem, we propose adding two new endpoints. One that is only

> used for binding to user's account, and another that is only for binding to

Reviewer 3: Again, this overview would be simpler to follow using the separate verbs "add" and "bind" for the actions that take place. Conveniently, these verbs also match the API paths, so that should make it even more straightforward.

an identity server of the user's choice. The existing binding endpoint will
be deprecated.

One may question why clients don't just contact an identity server directly
to bind a threepid, bypassing the implications of binding through a
homeserver. While this will work, binds should still occur through a
homeserver such that the homeserver can keep track of which binds were made,
which is important when a user wishes to deactivate their account (and remove
all of their bindings made on different identity servers).

A bind could be made on an identity server, which could then tell the
homeserver that a validation occured, but then there are security
considerations about how to authenticate an identity server in that instance
(and prevent people pretending to be identity servers and telling homeservers
about hundreds of fake binds to a user's account).

This MSC obseletes
[MSC2229](https://github.com/matrix-org/matrix-doc/pull/2229), which dealt
with changing the rules of the `bind` flag on the original endpoint. Since
that endpoint is being deprecated, the MSC is no longer relevant.

## Proposal

> Two new endpoints will be added to the Client Server API: `POST

Reviewer 4: I think the proposal would be much easier to grok if the first paragraph explained what the two endpoints actually do.  So `/bind` publishes the 3PID mapping on an IS, and `/add` adds an administrative 3PID to the HS?

Author: Yep, and ok can do. I suppose that is the idea of [BLUF](https://news.ycombinator.com/item?id=20964907).

minus the `bind` flag. The request parameters of `POST /account/3pid/add`
will simply consist of a JSON body containing `client_secret` and `sid`.

An example of binding a threepid to **an identity server only** with this new
endpoint is as follows:

First the client must request the threepid be validated by its chosen identity server.

```
POST https://identity.server/_matrix/identity/v2/validate/email/requestToken

{
    "client_secret": "don'tT3ll",
    "email": "bob@example.com",
    "send_attempt": 1
}
```

Once an email has been sent, the user clicks the link in the email, which
notifies the identity server that the email has been verified.

Next, the client completes the bind by calling the new endpoint on the homeserver:

```
POST https://home.server/_matrix/client/r0/account/3pid/bind

{
    "three_pid_creds": {
        "id_server": "example.org",
        "id_access_token": "abc123_OpaqueString",
        "sid": "abc123987",
        "client_secret": "don'tT3ll"
    }
}
```

The homeserver will then make a bind request to the specified identity server
on behalf of the user. The homeserver will record if the bind was successful
and notify the user.

> The threepid has now been binded on the user's requested identity server

Reviewer 4: ```suggestion
The threepid has now been bound on the user's requested identity server
```

without causing that threepid to be used for password resets or any other
homeserver-related functions.

For completeness, here is an example of binding a threepid to the
homeserver only, using the old endpoint:

The homeserver is validating the threepid in this instance, so the client
must use the `/requestToken` endpoint of the homeserver:

```
POST https://home.server/_matrix/client/r0/account/3pid/email/requestToken

{
    "client_secret": "don'tT3ll",
    "email": "bob@example.com",
    "send_attempt": 1,
}
```

Once an email has been sent, the user clicks the link in the email, which
notifies the homeserver that the threepid has been verified.

> The client then sends a request to the old endpoint on the homeserver to bind

Reviewer 4: why do we call this 'the old endpoint'?

Author: As we're deprecating it, but I can be more explicit.

Author: Typo, should be 'other'.

the threepid to user's account.

```
POST https://home.server/_matrix/client/r0/account/3pid/add

{
    "sid": "abc123987",
    "client_secret": "don'tT3ll"
}
```

The threepid will then be bound to the user's account.

> The achieve the above flows, some changes need to be made to existing

Reviewer 4: To achieve

# Separate Endpoints for Binding Threepids

> *Note: This MSC obseletes

Reviewer 5: ```suggestion
*Note: This MSC obsoletes
```

brought to the top level of the request body. The request parameters of `POST
/account/3pid/add` will simply consist of a JSON body containing
`client_secret` and `sid`.

The homeserver should prevent a threepid being added to a user's account if
it already part of another user's account. However, the homeserver should not
check for existing threepids when binding to an identity server. Identity
servers do not enforce this requirement and neither should the proxying
homeserver.

An example of binding a threepid to an identity server with this new endpoint
is as follows:

First the client must request the threepid be validated by its chosen identity server.

```
POST https://identity.server/_matrix/identity/v2/validate/email/requestToken

{
    "client_secret": "don'tT3ll",
    "email": "bob@example.com",
    "send_attempt": 1
}
```

The identity server must send an email to the specified address, including a
link to a URL on the identity server which will accept the validation session
ID, the given client_secret, and a randomly-generated token.

Once an email has been sent, the user clicks the link in the email, which
notifies the identity server that the email has been verified.

> Next, the client completes the bind by calling the new endpoint on the homeserver:

Reviewer 6: That "next" thing can (still) be triggered by `next_link` URL, right? Or am I outdated and we have a new way of passing the baton from IS back to the client application without awkward "Click here once you confirmed the email" buttons?

Author: No, we still have `next_link` shenanigans. It's specified in the `/requestToken` step, and homeservers should redirect users if it was specified as part of the validation session.

I can add a note for that.

Author: Actually, that part is handled on `submit_token`, which is covered by [MSC2263](https://github.com/matrix-org/matrix-doc/pull/2263). This MSC only deals with changing the `/account/3pid` endpoint, so I don't think `next_link` needs to be mentioned.

Now, consider the fact that the identity server used in this process is
provided by the user, using the endpoint's `id_server` parameter. If the user were
to supply a malicious identity server that would immediately answer "yes" to
any threepid validation, then the user could add any threepid to their
account on the homeserver (which is likely not something homeserver admins want).

To be clear, this is not a long-standing security issue. It is not a problem
in any released version of Synapse, as Synapse keeps a list of "trusted
identity servers" that acts a whitelist for what identity servers a user can
specify.

The concept of this whitelist is being removed in this MSC however, as part
of lessening the reliance of homeservers on identity servers. This cannot be
done while the homeserver is still trusting an identity server for validation
of threepids. If the endpoints are split, the homeserver will handle the
validation of threepids being added to user accounts, and identity servers
will validate threepids being bound to themselves.

## Proposal

To solve this problem, two new endpoints will be added to the Client Server

> API: `POST /account/3pid/bind` and `POST /account/3pid/add`. Binding to an

Reviewer 7: Its possibly a bit late in the day, but was there any discussion about using slightly more informative names? `POST /account/3pid/bind_to_is` or something? I had to look up the difference between `bind` and `add` when skimming this.

Author: We've used the `bind` vs. `add` terminology quite extensively for this project and it seems to have gone down well as soon as you're aware of the difference. So perhaps it's just a case of defining the terms in the spec?

The threepid has now been added to the user's account.

To achieve the above flows, some changes need to be made to existing

> endpoints. The `id_server` and `id_access_token` parameters are to be removed

Reviewer 2: removed or made optional? How does one proxy the calls to an IS if the fields aren't present?

Author: Err, yes, made optional as the spec still allows the homeserver to proxy requests to identity servers specified by the client.

Author: Actually, these parameters should be removed as it's dangerous to remove the requirement for a trusted identity server list while still allowing a user-provided `id_server` parameter. Really we should be removing `id_server` and `id_access_token` from all `/requestToken` endpoints. I'll add that in.

user's account.

To achieve the above flows, some changes need to be made to existing
endpoints. The [POST
/account/3pid](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-3pid)
endpoint is deprecated as the two new endpoints replace its functionality.
The `bind` parameter is to be removed, with the endpoint functioning as if
`bind` was `false`. Allowing an endpoint to add a threepid to both the
identity server and homeserver at the same time requires one to trust the
other, which is the exact behaviour we're trying to eliminate. Doing this
also helps backward compatibility, as explained in [Backwards
compatibility](#backwards-compatibility).

> The `id_server` and `id_access_token` parameters are to be removed

Reviewer 2: So this obsoletes [MSC2263](https://github.com/matrix-org/matrix-doc/pull/2263) too then? :/

They were made optional in 2263 for the reason of maintaining backwards compatibility with clients. Removing them does not maintain that backwards compatibility. In fact, 2263 even introduces an error code to let the HS make the decision about whether or not it cares through introduction of an error code.

This is duplicated effort.

Author: We can leave them as optional parameters but homeserver should ignore them else they are opening themselves up to malicious identity servers.

And then at that point there's not much point in keeping them as optional parameters versus removing them.

Reviewer 2: By removing them, it becomes impossible for `/bind` to work. The client will have no way of contacting an identity server through the homeserver, so the homeserver might end up rejecting `/bind`. If there's no way to proxy the calls to the IS through the HS, there's no point in `/bind` and we should just remove the `bind` flag from `/account/3pid` instead of splitting the endpoint.

I'd rather we keep the fields as optional and put a red box saying "you should abuse M_SERVER_NOT_TRUSTED at every given opportunity" until the deprecation period has passed and they get removed (as they are deprecated by MSC2263).

Reviewer 8: So, iiuc the problem is:
 * We don't want to support `id_server` in synapse because we want to ditch the concept of trusted ID servers
 * This isn't compatible with MSC2263 because that requires that the endpoints still work when given `id_server`

I think, as long as the requests still work in synapse with an `id_server` (modulo the fact they won't bind, but that's covered by this MSC), that's fine? The HS will send a token (not via the IS specified) but the right thing will still happen.

So I would suggest maybe keeping them optional for now?

Author: Keeping them optional but returning `M_SERVER_NOT_TRUSTED` works I suppose.

Reviewer 2: 2263 was specifically written to allow Synapse to not support `id_server` ftr: it can just throw `M_SERVER_NOT_TRUSTED` for everything.

Author: We can't have homeservers throw `M_SERVER_NOT_TRUSTED` else old clients trying to send `id_server` will get confused when they get this error.

But new servers can just ignore the parameter.

Reviewer 2: I'd still expect that servers throw the error instead of ignoring it, as ignoring it is against the spec.

It'd be no different than a 500 error in the eyes of the client (which it should be handling because `/requestToken` is certainly in the top 10 for most error-prone endpoints).

Reviewer 3: From a client perspective, we'd like old Riots to still be able add 3PIDs to the HS account. 

If `requestToken` begins throwing an error when `id_server` is included, those clients will no longer have any path to add a 3PID. The old Riots would have had known to special case and ignore `M_SERVER_NOT_TRUSTED`, which did not exist in the past.

Reviewer 2: @jryans sounds like the server shouldn't be throwing the error then and instead handle the `id_server` as deprecated (as proposed by #2263). If the server doesn't want to handle it, it is no different from it throwing a 500 error for a misconfigured email server.

It would be intentional that adding 3PIDs be blocked - it's an error related to something the server is not comfortable with doing.

This all should have come out on MSC2263, not here, as well. I'd still have the same opinion, but it's off-topic for this MSC (now).

* [POST /account/3pid/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-3pid-msisdn-requesttoken)
* [POST /register/email/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-register-email-requesttoken)
* [POST /register/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-register-msisdn-requesttoken)
* [POST /account/password/email/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-password-email-requesttoken)
* [POST /account/password/msisdn/requestToken](https://matrix.org/docs/spec/client_server/r0.5.0#post-matrix-client-r0-account-password-msisdn-requesttoken)

Either the homeserver itself or a service that the homeserver delegates to
should be handling the sending of validation messages, not a user-provided
server. Any mention of the homeserver being able to proxy to an identity
server in the above endpoint descriptions, as well as the text "It is
imperative that the homeserver keep a list of trusted Identity Servers and
only proxies to those that it trusts." is to be removed from all parts of the

> spec, as the homeserver should no longer need to trust any identity servers.

Reviewer 2: There's still an element of trust, as highlighted by yourself here: https://github.com/matrix-org/matrix-doc/pull/2263/files#r319536914

Author: Mm, Homeserver can get around this in implementation though (as we have done in Synapse) by blocking attempts to access private IP address ranges.

Reviewer 2: to be fair, keeping a list of servers that *aren't* trusted is by proxy keeping a list of servers that *are* trusted. However, yea: this is a bit of a pointless argument

Reducing the homeserver's trust in identity servers should be a boost to
security and improve decentralisation in the Matrix ecosystem to boot.

Some endpoints of the Client Server API allow a user to provide an
`id_server` parameter. Caution should be taken for homeserver developers to

> stop using these user-provided identity servers for any sensitive tasks, such

Reviewer 2: where possible, presumably. If it is flagged as required, there's not much the homeserver can do. If it is supposed to be optional in more places, please open another MSC.

Author: Yeah, where possible.
